---
title: <p style="font-size:45px;">mNET-seq</p>
date: "`r format(Sys.Date(), '%B %d, %Y')`"
output: 
  html_document:
    toc:            true
    toc_float:      true
    toc_depth:      3
    df_print:       "paged"
    code_folding:   "hide"
    self_contained: true
    theme:          "cosmo"
    highlight:      "tango"

params:
  config:            "configs/plots.yaml"
  proj:              "NET-seq"                                                            # Project name used for QC files
  overwrite:         false                                                                # Overwrite saved R objects
  template_dir:      "src"                                                                # Directory containing Rmd templates
  res_dir:           "results"                                                            # Results directory
  obj_dir:           "results/objects"                                                    # Directory to save R objects
  tab_dir:           "results/tables"                                                     # Directory to save tables
  ref_dir:           "~/ref"                                                              # Reference directory
  list_dir:          "gene_lists/GRCh38"                                                  # Gene list directory from ref_dir path
  genome:            "genome/GRCh38/Homo_sapiens.GRCh38.dna.primary_assembly.UCSC.fa.fai" # Chromosome sizes
  fasta:             "genome/GRCh38/Homo_sapiens.GRCh38.dna.primary_assembly.UCSC.fa"     # Fasta for pulling sequences
  genes_all:         "base_lists/GRCh38_HCT116_genes.bed.gz"
  genes_5:           "base_lists/GRCh38_5ksep.bed.gz"                                     # Gene list for 5' metaplots
  genes_3:           "base_lists/GRCh38_5ksep.bed.gz"                                     # Gene list for 3' metaplots
  genes_zone:        "base_lists/GRCh38_2ksep.bed.gz"                                     # Gene list for pausing zone plots
  genes_pause:       "base_lists/GRCh38_5ksep.bed.gz"                                     # Gene list for pause metaplots and sequence logos
  
  exons:             "/splicing_lists/GRCh38_exons.bed.gz"                                # Bed file containing exon coordinates
  trxn_info:         "/base_lists/base_gene_files/GRCh38_transcripts.bed.gz"              # Bed file containing transcript info to use for filtering exons
  
  gene_min_pct:      0.1                                                                  # Fraction of lowest expressing genes to exclude
  pause_strength:    "_strong_"                                                           # Pause strength label to use for loading input files
  pause_win:         200                                                                  # Pause window label to use for loading input files
  subsample_regions: ["gene", "tss", "body"]
  
  pause_lims:        [0, 1]                                                               # Min pauses for filtering TSS boxplots (1st value) and scatter plots (2nd value)
  tss_lim:           1                                                                    # Min number of raw reads for TSS boxplots
  body_lim:          1                                                                    # Min number of reads/kb for gene body region
  only_shared:       true                                                                 # Only include genes in TSS boxplots that have data for all regions
---

---

```{r "setup", echo = FALSE}

# Knit setup Rmd
knitr::knit(here::here(params$template_dir, "setup.Rmd"), "")

# Directories
res_dir <- here(params$res_dir)

# Check sample parameters
sections <- names(plot_grps$samples)
colors   <- names(plot_grps$colors)

if (!all(sections %in% colors)) {
  need <- sections[!sections %in% colors]
  
  stop("Colors were not provided for all samples: ", str_c(need, collapse = ", "))
}

# Create data.frame with sample info
sample_df <- plot_grps$samples %>%
  imap_dfr(~ {
    plt_grp <- .y
    clrs    <- plot_grps$colors[[plt_grp]]
    
    imap_dfr(.x, ~ {
      grp <- .y
      
      sam_clrs <- set_names(clrs, names(.x))
      
      imap_dfr(.x, ~ {
        sam <- .y
        
        map_dfr(.x, ~ {
          file <- .x
          
          tibble(
            sampling_grp = grp,
            plot_grp     = plt_grp,
            sample       = sam,
            file         = file,
            clr          = sam_clrs[[sam]]
          )
        })
      })
    })
  })

sample_df <- sample_df %>%
  separate(sample, into = c("sam", "rep"), sep = "_", remove = FALSE)

# Columns for bed files
bed_cols <- c(
  "chrom", "start", 
  "end",   "name", 
  "score", "strand"
)

win_cols <- c(bed_cols, "counts")
win_cols[5] <- "win_id"

# Genome files
genome <- read_tsv(
  file.path(params$ref_dir, params$genome),
  col_names = c("chrom", "size")
)

fa <- here(params$ref_dir, params$fasta)

# Gene lists and filter based on length
load_genes <- function(drctry, file, len = 1, rm_cols = TRUE) {
  res <- here(drctry, file) %>%
    read_tsv(col_names = bed_cols) %>%
    mutate(gene_len = (end - start) / 1000) %>%
    filter(gene_len > len)
  
  if (rm_cols) {
    res <- res %>%
      select(name, gene_len)
  }
  
  res
}

list_dir <- file.path(params$ref_dir, params$list_dir)

genes_5     <- load_genes(list_dir, params$genes_5, 2)
genes_3     <- load_genes(list_dir, params$genes_3, 2)
genes_pause <- load_genes(list_dir, params$genes_pause, 2)

genes_exons <- load_genes(
  drctry  = list_dir,
  file    = params$genes_zone,
  len     = 4,
  rm_cols = FALSE
)

genes_zone  <- genes_exons %>%
  select(name, gene_len)

genes_exons <- genes_exons %>%
  select(1:6)

```

```{r "load exons"}

ex_cols <- c(
  bed_cols,    "symbol",
  "gene_id",   "gene_name",
  "txn_id",    "ex_coords",
  "in_coords", "ex_len",
  "in_len",    "biotype",
  "seq"
)

exons <- read_tsv(
  str_c(list_dir, params$exons),
  col_names = ex_cols
)

# Transcript info
# This must be passed separately since the exon bed file is generated from the
# gene pred file. There is no easy way to add the transcript_support_level to
# the gene pred file.
tx_cols <- c(
  bed_cols, "gene_id",
  "symbol", "biotype",
  "txn_id", "support"
)
  
tx_info <- here(list_dir, params$trxn_info) %>%
  read_tsv(col_names = tx_cols) %>%
  select(txn_id, support)

```

```{r "1st exon length"}

# Identify first exon coordinates
# join with gene list to get gene coordinates
# only keep genes with 1st exon matching the gene TSS coordinate
exon_lens <- exons %>%
  left_join(genes_exons, by = c(gene_name = "name"), suffix = c("", ".gene")) %>%
  filter(
    !is.na(chrom.gene),
    score == 1,          # exon number, only keep 1st exon
    ((strand == "+" & start == start.gene) | (strand == "-" & end == end.gene))
  ) %>%
  
  # Add transcript support level
  left_join(tx_info, by = "txn_id") %>%
  
  select(
    chrom, start, end, score,
    strand, gene_name, txn_id, ex_coords,
    in_coords, ex_len, support
  ) %>%
  
  # Remove single exon genes
  # this is ~470 genes
  filter(!is.na(in_coords) & !is.na(support)) %>%
  
  # Use highest confidence transcript
  group_by(gene_name) %>%
  filter(support == min(support)) %>%
  ungroup()

# Transcripts to use for 5'ss metaplots
# some genes will have multiple transcripts
high_conf_txns <- unique(exon_lens$txn_id)

# Transcripts to use for quantifying exon/intron pausing
# exclude genes with multiple transcript IDs (~500)
ex_int_txns <- exon_lens %>%
  group_by(gene_name) %>%
  filter(n_distinct(txn_id) == 1) %>%
  pull(txn_id)

# Remove genes with multiple 1st exon coordinates
# this is a small number (~100)
exon_lens <- exon_lens %>%
  group_by(gene_name) %>%
  filter(n_distinct(ex_coords) == 1) %>%
  group_by(ex_len, .add = TRUE) %>%
  summarize(
    txn_id = first(txn_id),
    .groups = "drop"
  )

```

```{r "plot exon length", fig.width = 6, fig.height = 3, include = FALSE}

lab <- paste0(
  "n = ", comma(nrow(exon_lens)),
  "\nmean: ", round(mean(exon_lens$ex_len), 0), " bp"
)

exon_lens %>%
  ggplot(aes(ex_len, ..count..)) +
  geom_histogram(fill = "#EC6E12", alpha = 0.75) +
  geom_vline(xintercept = mean(exon_lens$ex_len), linetype = 2) +
  
  annotate(
    geom  = "text",
    label = lab,
    x     = 10000,
    y     = 1100,
    hjust = 1
  ) +
  labs(x = "1st exon length", y = "Number of genes") +
  scale_x_log10() +
  theme_info

```

# Quality Control

`r knitr::knit_child(here(params$template_dir, "qc-template.Rmd"))`

```{r "build chunks"}

# Create list of chunks
template <- here(params$template_dir, "analysis-template.Rmd")

chunks <- names(plot_grps$samples) %>%
  map(~ knit_expand(template))

```

`r knit_child(text = chunks)`

# Example genes

```{r "ENSEMBL DATABASE"}

# Gene coordinates
genes_all <- read_tsv(
  file.path(params$ref_dir, params$list_dir, params$genes_all),
  col_names = bed_cols
)

# Create database from gtf
gtf <- "/home/rmsheridan/ref/annotation/GRCh38/gtf/Homo_sapiens.GRCh38.95.gtf"

ensdb <- here("ref/Homo_sapiens.GRCh38.95.sqlite")

if (!file.exists(ensdb)) {
  ensdb <- ensDbFromGtf(gtf = gtf, outfile = ensdb)
}

db <- EnsDb(ensdb)

```

```{r "FIG 1C", fig.width = 6, fig.height = 5}

# Load data for browser plots
bg_sams     <- c("+Spt5_r1", "+Spt5_r2")
bg_grps     <- c("Spt5-r1", "Spt5-r2")
bg_plt_grps <- c("Spt5", "Spt5")

bgs     <- .get_bgs(bg_sams, bg_grps, bg_plt_grps, sample_df)
bg_clrs <- .get_bg_clrs(bg_sams, bg_plt_grps, sample_df)
pauses  <- .get_pauses(bg_sams, bg_grps, bg_plt_grps, sample_df)
pzones  <- .get_zones(bg_plt_grps, genes_all)

gene <- "BCAM"

create_browser(
  bgs        = bgs,
  gene       = gene,
  genes_df   = genes_all,
  genes_db   = db,
  pauses     = pauses,
  zones      = pzones,
  flank      = c(0.08, 0),
  track_clrs = bg_clrs,
  track_ttls = c("rep 1", "rep 2"),
  equal_axis = TRUE,
  pause_track_height = 0.1,
  gene_track_height  = 0.1,
  fontsize   = 18
)

```

<br>

```{r "FIG 4G", fig.width = 8, fig.height = 7}

# Load data for browser plots
bg_sams     <- c("+Spt5_r1", "-Spt5_r1", "+Spt6_r1", "-Spt6_r1")
bg_grps     <- c("Spt5-r1", "Spt5-r1", "Spt6-r1", "Spt6-r1")
bg_plt_grps <- c("Spt5", "Spt5", "Spt6", "Spt6")

bgs     <- .get_bgs(bg_sams, bg_grps, bg_plt_grps, sample_df)
bg_clrs <- .get_bg_clrs(bg_sams, bg_plt_grps, sample_df)
pauses  <- .get_pauses(bg_sams, bg_grps, bg_plt_grps, sample_df)

gene <- "FOS"

create_browser(
  bgs        = bgs,
  gene       = gene,
  genes_df   = genes_all,
  genes_db   = db,
  pauses     = pauses,
  flank      = c(0.05, -0.5),
  track_clrs = bg_clrs,
  equal_axis = TRUE,
  track_ttls = rep(" ", length(bg_sams)),
  pause_track_height = 0.1,
  gene_track_height  = 0.1,
  y_dec      = 0,
  fontsize   = 18
)

```

<br>

```{r "FIG 4H", fig.width = 8, fig.height = 4}

# Load data for browser plots
bg_sams     <- c("+Spt5_r1", "-Spt5_r1")
bg_grps     <- c("Spt5-r1", "Spt5-r1")
bg_plt_grps <- c("Spt5", "Spt5")

bgs     <- .get_bgs(bg_sams, bg_grps, bg_plt_grps, sample_df)
bg_clrs <- .get_bg_clrs(bg_sams, bg_plt_grps, sample_df)
pauses  <- .get_pauses(bg_sams, bg_grps, bg_plt_grps, sample_df)

gene <- "ACTB"

create_browser(
  bgs         = bgs,
  gene        = gene,
  genes_df    = genes_all,
  genes_db    = db,
  pauses      = pauses,
  flank       = c(-0.985, 0),
  track_clrs  = bg_clrs,
  equal_axis  = TRUE,
  track_ttls  = rep(" ", length(bg_sams)),
  line_width  = 0,
  y_dec       = 0,
  fontsize    = 18,
  track_scale = 0.5,
  pause_track_height = 0.1,
  gene_track_height  = 0.1
)

```

<br>

```{r "FIG 6A", fig.width = 6, fig.height = 7}

# Load data for browser plots
bg_sams     <- c("WT_r1",  "WT_r2",  "SA3_r1", "SA3_r2", "SE3_r1", "SE3_r2")
bg_grps     <- c("Spt5-sa-r1", "Spt5-sa-r2", "Spt5-sa-r1", "Spt5-sa-r2", "Spt5-se-r1", "Spt5-se-r2")
bg_plt_grps <- c("Spt5-SA", "Spt5-SA", "Spt5-SA", "Spt5-SA", "Spt5-SE", "Spt5-SE")

bgs     <- .get_bgs(bg_sams, bg_grps, bg_plt_grps, sample_df)
bg_clrs <- .get_bg_clrs(bg_sams, bg_plt_grps, sample_df)
pauses  <- .get_pauses(bg_sams, bg_grps, bg_plt_grps, sample_df)

gene <- "PBDC1"

create_browser(
  bgs         = bgs,
  gene        = gene,
  genes_df    = genes_all,
  genes_db    = db,
  pauses      = pauses,
  flank       = c(0.01, -0.95),
  track_clrs  = bg_clrs,
  track_ttls  = rep(" ", length(bg_sams)),
  line_width  = 2,
  y_dec       = 1,
  fontsize    = 18,
  track_scale = 0.2,
  pause_track_height = 0.15,
  gene_track_height  = 0.2,
  scale_track_height = 0.15
)

```

<br>

```{r "FIG S2F", fig.width = 6, fig.height = 7, eval = FALSE}

# Load data for browser plots
bg_sams     <- c("WT_r1",  "WT_r2",  "SA3_r1", "SA3_r2", "SE3_r1", "SE3_r2")
bg_grps     <- c("Spt5-sa-r1", "Spt5-sa-r2", "Spt5-sa-r1", "Spt5-sa-r2", "Spt5-se-r1", "Spt5-se-r2")
bg_plt_grps <- c("Spt5-SA", "Spt5-SA", "Spt5-SA", "Spt5-SA", "Spt5-SE", "Spt5-SE")

bgs     <- .get_bgs(bg_sams, bg_grps, bg_plt_grps, sample_df)
bg_clrs <- .get_bg_clrs(bg_sams, bg_plt_grps, sample_df)
pauses  <- .get_pauses(bg_sams, bg_grps, bg_plt_grps, sample_df)

gene <- "PBDC1"

create_browser(
  bgs         = bgs,
  gene        = gene,
  genes_df    = genes_all,
  genes_db    = db,
  pauses      = pauses,
  flank       = c(0.01, -0.95),
  track_clrs  = bg_clrs,
  track_ttls  = rep(" ", length(bg_sams)),
  line_width  = 2,
  y_dec       = 1,
  fontsize    = 18,
  track_scale = 0.2,
  pause_track_height = 0.15,
  gene_track_height  = 0.2,
  scale_track_height = 0.15
)

```

