
```{r "{{.x}} sample names", cache = FALSE}

# Strings to match samples and create labels
grp     <- "{{.x}}"
samples <- plot_grps$samples[[grp]]
clrs    <- plot_grps$colors[[grp]]

win   <- params$pause_win
stren <- params$pause_strength

pause_prfx <- str_c("_", win, stren)

# Sample names
# the names in the samples list should correspond to the sampling group
sam_df <- sample_df %>%
  filter(plot_grp == grp)

sam_nms <- set_names(
  sam_df$sample,
  sam_df$file
)

sam_lnms <- set_names(
  sam_df$sample,
  str_c(sam_df$file, "-", sam_df$sampling_grp)
)

sams <- names(sam_lnms)

con_sams <- samples %>%
  map(pluck, 1) %>%
  imap(~ str_c(.x, "-", .y)) %>%
  unlist(use.names = FALSE)

con_sams <- sam_lnms[con_sams]

# Subsample sample names
sub_regs <- set_names(params$subsample_regions)

sub_sams <- sub_regs %>%
  map(~ str_c(sams, "_", .x))

sub_sam_lnms <- sub_sams %>%
  map(~ set_names(sam_df$sample, .x))

sub_con_sams <- sub_regs %>%
  map(~ {
    reg <- .x
    
    samples %>%
      map(pluck, 1) %>%
      imap(~ str_c(.x, "-", .y, "_", reg)) %>%
      unlist(use.names = FALSE)
  })

sub_con_sams <- sub_sam_lnms %>%
  imap(~ .x[sub_con_sams[[.y]]])

# Treatments
n_sams <- length(sam_nms)

treat_reps <- sam_nms %>%
  str_split("_")

treats <- treat_reps %>%
  map_chr(pluck, 1) %>%
  unique()

n_treats <- length(treats)

if (n_treats != length(clrs)) {
  stop("Must provide a color for each sample: ", grp)
}

# Replicates
reps <- treat_reps %>%
  map_chr(pluck, 2) %>%
  unique()

n_reps <- length(reps)

common_reps <- sam_df %>%
  group_by(rep) %>%
  filter(all(treats %in% sam)) %>%
  pull(rep) %>%
  unique()

n_common_reps <- length(common_reps)

# Treatment groups to use when calculating fold changes
fc_treats <- c(treats[n_treats - 1], last(treats))
  
if (identical(n_treats, 1L)) {
  fc_treats <- rep(fc_treats, 2)
}

fc_lab <- str_c(fc_treats[2], "/", fc_treats[1])

# Sample colors
clrs <- set_names(clrs, treats)

# Sample directories
pause_dir <- set_names(sams) %>%
  map_chr(~ here(res_dir, .x, "pauses"))

meta_dir <- set_names(sams) %>%
  map_chr(~ here(res_dir, .x, "metaplot_beds"))

# Subsample sample directories
get_file_dir <- function(sam_dir, file_dir) {
  sam_dir %>%
    set_names() %>%
    map_chr(~ here(res_dir, .x, file_dir))
}

sub_pause_dir <- sub_sams %>%
  map(get_file_dir, "pauses")

sub_meta_dir <- sub_sams %>%
  map(get_file_dir, "metaplot_beds")

```

```{r "{{.x}} chunks", cache = FALSE}

templates <- c(
  motif = "pause-motif-template.Rmd",
  quant = "pause-quant-template.Rmd",
  zone  = "pause-zone-template.Rmd"
) %>%
  map_chr(~ here(params$template_dir, .x))

chunks <- templates %>%
  map(~ {
    knit_expand(
      file  = .x,
      grp   = grp,
      stren = stren,
      win   = win
    )
  })

```

```{r "{{.x}} gene lists"}

# Samples to use signal for filtering genes
# for WT use all samples
g_sams <- con_sams

if (grp == "WT") {
  g_sams <- sam_lnms
}

# Load and merge bed files
paths <- g_sams %>%
  imap(~ {
    p <- here(res_dir, .y, "metaplot_beds/")
    p <- str_c(p, str_c(.y, "_gene_S_N.bed.gz"))
    
    names(p) <- .x
    p
  }) %>%
  purrr::reduce(c)

gene_coords <- paths %>%
  imap(~ {
    .x %>%
      load_bed(c(bed_cols, "counts")) %>%
      mutate(sample = .y) %>%
      separate(sample, sep = "_", into = c("treat", "rep"), remove = FALSE)
  }) %>%
  merge_beds(bind = TRUE) %>%
  mutate(gene_len = (end - start) / 1000) %>%
  group_by(chrom, start, end, name, strand, gene_len) %>%
  summarize(counts = mean(counts), .groups = "drop")

# Filter genes to remove poorly expressed genes
# genes must have >0 signal and not be in the bottom params$gene_min_pct of
# genes based on signal
genes <- gene_coords %>%
  mutate(pct = percent_rank(counts)) %>%
  filter(counts > 0 & pct > params$gene_min_pct) %>%
  dplyr::select(name, counts)

g_5 <- genes_5 %>%
  semi_join(genes, by = "name")

g_3 <- genes_3 %>%
  semi_join(genes, by = "name")

g_z <- genes_zone %>%
  semi_join(genes, by = "name")

g_p <- genes_pause %>%
  semi_join(genes, by = "name")

```

# `r grp`

```{r "{{.x}} sample table"}

sam_df %>%
  select(-plot_grp)
  
```

Mean mNET-seq signal (top) or mean relative signal (bottom) is shown below for 5' (10 bp bins) and 3' regions (50 bp bins).

* genes >2 kb long and separated by >5 kb
* bottom `r params$gene_min_pct * 100`% of genes were removed

```{r "{{.x}} 5' mean"}

# Load and merge bed files
sfxs <- c(
  "sense"      = "_5-10bp_S_N.bed.gz",
  "anti-sense" = "_5-10bp_AS_N.bed.gz"
)

out <- here(params$obj_dir, str_c(grp, "_NET_5_10bp.tsv.gz"))

merge_5 <- load_merge_wins(
  prfxs     = names(meta_dir),   # Sample names
  sfxs      = sfxs,              # Suffix list with types as names
  paths     = meta_dir,          # Directories containing files
  group     = grp,               # Group name
  genes     = g_5,               # Genes to filter
  
  file_out  = out,               # Path to output file
  overwrite = params$overwrite,  # Overwrite output file if it exists
  col_names = win_cols,          # Column names for bed files
  win_num   = 800,               # Total number of expected windows (including sense + anti)
  ref_win   = 201,               # Reference window for adjusting window ids
  win_min   = 101                # Min window for filtering
) %>%
  check_group_size(
    clmns = c("name", "sample", "type", "win_id"),
    quiet = TRUE
  )

# Calculate mean signal
mean_5 <- merge_5 %>%
  calc_mean_signal(
    rel  = FALSE,
    grps = c("sample", "group", "type")
  ) %>%
  format_sample_names(add_grp = TRUE)

```

```{r "{{.x}} 3' mean"}

# Load bed files
sfxs <- c(
  "sense"      = "_3-50bp_S_N.bed.gz",
  "anti-sense" = "_3-50bp_AS_N.bed.gz"
)

out <- here(params$obj_dir, str_c(grp, "_NET_3_50bp.tsv.gz"))

merge_3 <- load_merge_wins(
  prfxs     = names(meta_dir),   # Sample list with group as name
  sfxs      = sfxs,              # Suffix list with types as names
  paths     = meta_dir,          # Directories containing files
  group     = grp,               # Group name
  genes     = g_3,               # Genes to filter
  
  file_out  = out,               # Path to output file
  overwrite = params$overwrite,  # Overwrite output file if it exists
  col_names = win_cols,          # Column names for bed files
  win_num   = 400,               # Total number of expected windows (including sense + anti)
  ref_win   = 100,               # Reference window for adjusting window ids
  win_min   = 61,                # Min window for filtering
  win_max   = 170                # Max window for filtering
) %>%
  check_group_size(
    clmns = c("name", "sample", "type", "win_id"),
    quiet = TRUE
  )
  
# Calculate mean signal
mean_3 <- merge_3 %>%
  calc_mean_signal(
    rel  = FALSE,
    grps = c("sample", "group", "type")
  ) %>%
  format_sample_names(add_grp = TRUE)

```

```{r "{{.x}} 5' 3' metaplots", fig.width = 9, fig.height = 4}

# Create final figure
ylim_3 <- set_strand_ratio(mean_3, mean_5, mtplyr = 4)

se <- NULL

if (n_reps > 1) {
  se <- "rep"
}

create_meta_fig(
  df_5      = mean_5,
  df_3      = mean_3,
  color     = "treat",
  ylim_3    = ylim_3,
  sams      = treats,
  plot_clrs = clrs,
  se_clmn   = se,
  plot_ttl  = "mean eNET-seq (RPKM)"
)

```

```{r "{{.x}} 5' 3' rel metaplots", fig.width = 10, fig.height = 4}

# Calculate mean signal
rel_5 <- merge_5 %>%
  calc_mean_signal(
    rel  = TRUE,
    grps = c("sample", "group", "type")
  ) %>%
  format_sample_names(add_grp = TRUE)

# Calculate mean signal
rel_3 <- merge_3 %>%
  calc_mean_signal(
    rel  = TRUE,
    grps = c("sample", "group", "type")
  ) %>%
  format_sample_names(add_grp = TRUE)

# Create final figure
ylim_3 <- set_strand_ratio(rel_3, rel_5, mtplyr = 4)

create_meta_fig(
  df_5      = rel_5,
  df_3      = rel_3,
  color     = "treat",
  ylim_3    = ylim_3,
  sams      = treats,
  plot_clrs = clrs,
  se_clmn   = se,
  plot_ttl  = "relative eNET-seq signal"
)

```

```{r "{{.x}} ANTISENSE TEST", eval = FALSE}

strand_class <- merge_3 %>%
  filter(
    win_dist > 0,
    win_dist < 3,
    sample == sams[1]
  ) %>%
  group_by(sample, name, type) %>%
  summarize(
    counts  = sum(counts),
    .groups = "drop"
  ) %>%
  pivot_wider(names_from = "type", values_from = "counts") %>%
  mutate(
    class = ifelse(sense > `anti-sense`, "sense", "anti"),
    rat   = `anti-sense` / sense
  ) %>%
  select(name, class, rat)

plt_dat <- merge_3 %>%
  left_join(strand_class, by = "name")

plt_dat %>%
  group_by(sample, type, class, win_dist) %>%
  summarize(counts = mean(counts), .groups = "drop") %>%
  mutate(counts = ifelse(type == "sense", counts, counts * -1)) %>%
  
  ggplot(aes(win_dist, counts, color = sample, alpha = type)) +
  geom_line() +
  scale_alpha_manual(values = c(0.5, 1)) +
  facet_wrap(~ class, scales = "free_y")

```

<br>

Zoomed-in versions of the 5' and 3' metaplots are shown below.

```{r "{{.x}} 5' 3' zoom metaplots", fig.width = 10, fig.height = 5}

meta_theme <- theme(
  aspect.ratio = 0.9,
  legend.position = "top"
)

if (n_treats == 1) {
  meta_theme$legend.position <- "none"
}

# Create zoomed 5' plots
z_dfs <- list(
  "mean eNET-seq (RPKM)"     = mean_5,
  "relative eNET-seq signal" = rel_5
)

z5_plts <- z_dfs %>%
  imap(~ {
    res <- .x %>%
      filter(win_dist >= -0.5 & win_dist <= 0.5) %>%
      create_meta(
        x         = "win_dist",
        y         = "counts",
        color     = "treat",
        alpha     = "type",
        n_lab     = "n",
        se_clmn   = se,
        plot_clrs = clrs,
        plot_lvls = treats
      ) +
      labs(y = .y) +
      add_breaks(seq(-0.6, 0.6, 0.2), "TSS") +
      meta_theme
    
    res
  })

# Create zoomed 3' plots
z_dfs <- list(
  "mean eNET-seq (RPKM)"     = mean_3,
  "relative eNET-seq signal" = rel_3
)

z3_plts <- z_dfs %>%
  imap(~ {
    res <- .x %>%
      create_meta(
        x         = "win_dist",
        y         = "counts",
        color     = "treat",
        alpha     = "type",
        n_lab     = "n",
        se_clmn   = se,
        plot_clrs = clrs,
        plot_lvls = treats
      ) +
      labs(y = .y) +
      add_breaks(seq(-3, 3, 1.5), "pAS") +
      meta_theme
    
    res
  })

# Creat final figures
z_plts <- map2(z5_plts, z3_plts, ~ {
  plot_grid(
    .x, .y,
    align = "vh",
    axis  = "trbl"
  )
})

z_plts[[1]]
z_plts[[2]]

```

```{r "{{.x}} 3' zoom metaplots", fig.width = 10, fig.height = 5}

z3_plts_2 <- z_dfs %>%
  imap(~ {
    res <- .x %>%
      filter(type == "sense") %>%
      create_meta(
        x          = "win_dist",
        y          = "counts",
        color      = "treat",
        alpha      = "type",
        n_lab      = "n",
        se_clmn    = se,
        plot_clrs  = clrs,
        plot_lvls  = treats,
        h_line     = NULL,
        plot_alphs = 1
      ) +
      labs(y = .y) +
      add_breaks(seq(-3, 3, 1.5), "pAS") +
      meta_theme +
      expand_limits(y = 0) +
      theme(
        aspect.ratio = 0.7
      )
    
    res
  })

z3_plts_2 %>%
  plot_grid(
    plotlist = .,
    nrow = 1
  )

```

<br>

Mean mNET-seq signal is shown below for the region around 5' splice sites.

```{r "{{.x}} ss functions", eval = n_treats > 1}

# Create metaplot bargraph
create_ss_bars <- function(df_in, y, color, yttl = waiver(), wins = c(41, 60),
                           ref_win = 51, plt_clrs = clrs) {
  
  fill_clrs    <- plt_clrs
  fill_clrs[1] <- "white"
  
  gd_aes <- list(size = 5, shape = 15, alpha = 0.5)
  gd_lgd <- guide_legend(override.aes = gd_aes)
  
  res <- df_in %>%
    convert_to_bars() %>%
    
    ggplot(aes(win_id, !!sym(y), color = !!sym(color), fill = !!sym(color))) +
    
    geom_vline(xintercept = 50.5, size = 0.5, linetype = 2) +

    geom_ribbon(
      aes(ymin = 0, ymax = !!sym(y)),
      alpha        = 0.05,
      size         = 0.25,
      outline.type = "upper",
      key_glyph    = draw_key_point
    ) +
    
    geom_text(
      aes(Inf, Inf, label = n),
      color = "black",
      hjust = 1.1,
      vjust = 1.4,
      size  = txt_pt / .pt,
      check_overlap = TRUE
    ) +
    
    guides(color = gd_lgd, fill = FALSE) +
    
    scale_color_manual(values = plt_clrs) +
    scale_fill_manual(values = fill_clrs) +
    
    theme_info +
    theme(
      legend.position  = "top",
      legend.title     = element_blank(),
      axis.title.x     = element_blank(),
      axis.text.x      = element_text(hjust = 0.7)
      # strip.background = element_rect(fill = "#fafafa", color = "grey85"),
    ) +
    labs(y = yttl)
  
  # Set x-axis breaks/labels
  brk_sz <- 10
  
  lab_fn <- function(x, ref = 51) {
    ifelse(
      x <= ref - 1,
      x - ref,
      str_c("+", x - (ref - 1))
    )
  }
  
  brk_fn <- function(wins, ref, sz = 10) {
    c(
      seq(wins[1], ref - 1, sz),
      seq(ref + sz - 1, wins[2], sz)
    )
  }
  
  brks  <- brk_fn(wins, ref = ref_win, sz = 10)
  xlabs <- lab_fn(brks, ref = ref_win)
  
  res <- res +
    scale_x_continuous(
      labels = xlabs,
      breaks = brks
    )
  
  res
}

# Add extra values so signal appears as bar graph
convert_to_bars <- function(df_in, xclmn = "win_id", yclmn = "counts") {
  res <- df_in %>%
    rowwise() %>%
    mutate(
      !!sym(xclmn) := list(c(!!sym(xclmn) - 0.5, !!sym(xclmn) + 0.5)),
      !!sym(yclmn) := list(rep(!!sym(yclmn), 2))
    ) %>%
    ungroup() %>%
    unnest(all_of(c(xclmn, yclmn)))
  
  res
}

```

```{r "{{.x}} meta exons", eval = n_treats > 1}

# Get exons to use for metaplots
# only include separated genes
# most genes not in high_conf_txns (545) are single exon genes
meta_exons <- exons %>%
  filter(
    gene_name %in% g_5$name,
    txn_id %in% high_conf_txns,
    ex_len > 50,
    in_len > 50
  ) %>%
  
  # Re-format exon name to include 5' ss coordinates
  # this is required so names are compatible with metaplot beds
  bed_flank(right = 1, genome = genome, strand = TRUE) %>%
  mutate(
    name = str_c(chrom, ":", start, "-", end),
    name = str_c(name, ";", gene_name, ";", seq)
  ) %>%
  
  group_by(name, gene_name) %>%
  summarize(
    score   = min(score, na.rm = TRUE),
    .groups = "drop"
  ) %>%
  
  # Collapse so 5' ss coordinates are not duplicated
  # it is possible that different exons have the same 5' ss coordinates
  # use the lowest exon number for unique coordinates
  mutate(
    class = if_else(score == 1, "first", "other"),
    .keep = "unused"
  )

```

```{r "{{.x}} ss data", eval = n_treats > 1}

# Load and merge bed files
sfxs <- c("sense" = "_5ss-wins_S_N.bed.gz")

out <- here(params$obj_dir, str_c(grp, "_NET_5ss.tsv.gz"))

merge_5ss <- load_merge_wins(
  prfxs     = names(meta_dir),   # Sample names
  sfxs      = sfxs,              # Suffix list with types as names
  paths     = meta_dir,          # Directories containing files
  group     = grp,               # Group name
  
  file_out  = out,               # Path to output file
  overwrite = params$overwrite,  # Overwrite output file if it exists
  col_names = win_cols,          # Column names for bed files
  win_num   = 100,
  win_len   = 1,
  genes     = meta_exons
) %>%
  check_group_size(
    clmns = c("name", "sample", "type", "win_id"),
    quiet = TRUE
  )

# Calculate mean signal
mean_5ss <- merge_5ss %>%
  left_join(meta_exons, by = "name") %>%
  filter(!is.na(class)) %>%
  select(sample, name, win_id, class, qrt, counts) %>%
  calc_mean_signal(
    rel       = FALSE,
    win_col   = "win_id",
    grps      = c("sample", "class"),
    flip_type = NULL
  ) %>%
  format_sample_names(add_grp = TRUE)

```

```{r "{{.x}} ss metaplot", fig.width = 10, fig.height = 1 + (2 * n_reps), eval = n_treats > 1}

# Create metaplots
mean_5ss %>%
  create_ss_bars(
    y        = "counts",
    color    = "treat",
    wins     = c(1, 100),
    plt_clrs = clrs,
    yttl     = "mean eNET-seq (RPKM)"
  ) +
  facet_wrap(rep ~ class, scale = "free_y")

```

<br>

Zoomed-in versions of 5' splice site metaplots for 1st exons are shown below. 

```{r "{{.x}} zoom ss metaplot", fig.width = 10, fig.height = 1 + (2 * n_reps), eval = n_treats > 1}

# Create metaplots
dat <- mean_5ss %>%
  filter(class == "first")

mx_sig <- dat %>%
  filter(win_id < 50) %>%
  pull(counts) %>%
  max()

dat %>%
  create_ss_bars(
    y        = "counts",
    color    = "treat",
    wins     = c(1, 100),
    plt_clrs = clrs,
    yttl     = "mean eNET-seq (RPKM)"
  ) +
  facet_wrap(rep ~ class, scale = "free_y") +
  coord_cartesian(ylim = c(0, mx_sig * 1.01))

```

<br>

## Pause motifs

`r knit_child(text = chunks$motif)`

<br>

## Pausing quantification

`r knit_child(text = chunks$quant)`

<br>

## Pausing zones

`r knit_child(text = chunks$zone)`

---

<br>

<br>

