
```{r "{{grp}} pause stats setup"}

grp   <- "{{grp}}"
stren <- "{{stren}}"
win   <- {{win}}

sep_prfx   <- str_c("_", win, "_sep", stren)
pause_prfx <- str_c("_", win, stren)

```

```{r "{{grp}} pause funs"}

# Helper to load pause bed files
load_pause_beds <- function(prfxs, sfxs, paths, genes = NULL, regions = NULL) {
  paths <- str_c(paths, "/", prfxs, sfxs)
  paths <- set_names(paths, prfxs)
  
  res <- paths %>%
    map(read_tsv, col_names = bed_cols)
  
  res <- res %>%
    imap_dfr(~ {
      .x <- .x %>%
        separate(
          name,
          into   = c("pause_coords", "gene_name"),
          sep    = "\\*",
          remove = FALSE
        ) %>%
        mutate(
          sample = .y,
          name   = paste0(name, "*", score)
        ) %>%
        select(-pause_coords)
      
      if (!is.null(genes)) {
        .x <- .x %>%
          semi_join(genes, by = c("gene_name" = "name"))
      }
      
      .x
      
      # THIS PRODUCES SPORADIC ERROR WHEN KNITTING: "embedded nul in string"
      # .x %>%
      #   mutate(
      #     sample    = .y,
      #     gene_name = str_extract(name, "(?<=\\*).+$"),
      #     name      = paste0(name, "*", score)
      #   ) %>%
      #   semi_join(genes, by = c("gene_name" = "name"))
    })

  n_pauses <- nrow(res)
  
  # Intersect pauses
  if (!is.null(regions)) {
    res <- regions %>%
      imap_dfr(~ {
        bed_intersect(res, .x, suffix = c("", ".y")) %>%
          filter(
            .overlap == 1,      # by default bookended intervals are included
            strand == strand.y  # some pAS regions from opposite strands will overlap between genes
          ) %>%
          group_by(sample) %>%
          mutate(
            region = .y,
            n_lab  = str_c(comma(n_distinct(name)), " pauses\n"),
            n_lab  = str_c(n_lab, comma(n_distinct(gene_name)), " genes"),
          ) %>%
          select(-ends_with(".y"), -.overlap)
      }) %>%
      ungroup()

    if (n_pauses != nrow(res)) {
      stop(
        "The number of pauses after intersection does not match the number of ",
        "input pauses, data.frame is malformed. This could be due to overlap ",
        "between the provided regions or overlap between genes."
      )
    }
  }

  res
}

# Calculate overlap between pause bed files
calc_pause_overlap <- function(dfs) {
  
  if (length(dfs) < 2) {
    return(NULL)
  }
  
  # Identify shared pauses
  y_i <- 2:length(dfs)
  
  shared <- bed_intersect(
    x      = dfs[[1]],
    y      = dfs[y_i],
    suffix = c("", ".y")
  ) %>%
    filter(
      strand == strand.y,
      .overlap > 0
    )
  
  # If more than two samples get pauses shared between all
  if (length(dfs) > 2) {
    n_y <- seq_along(y_i)
    
    shared <- shared %>%
      group_by(name) %>%
      filter(all(n_y %in% .source)) %>%
      ungroup()
  }
  
  shared <- shared %>%
    select(-ends_with(".y"), -starts_with(".")) %>%
    distinct()
  
  n_shared <- nrow(shared)
  
  res <- tibble(
    key      = str_c(names(dfs), collapse = " "),
    class    = "shared",
    n_pauses = n_shared,
    n_genes  = n_distinct(shared$gene_name),
    n        = n_pauses
  )
  
  # Identify unique pauses
  uniq <- tibble(
    key      = names(dfs),
    class    = "unique",
    n_pauses = map_int(dfs, nrow),
    n_genes  = map_int(dfs, ~ n_distinct(.x$gene_name)),
    n        = n_pauses - n_shared
  )
  
  # Create results table
  res <- bind_rows(res, uniq) %>%
    mutate(frac = n / sum(n))
  
  res
}

# Add extra values so signal appears as bar graph
convert_to_bars <- function(df_in, xclmn = "win_id", yclmn = "counts") {
  res <- df_in %>%
    rowwise() %>%
    mutate(
      !!sym(xclmn) := list(c(!!sym(xclmn) - 0.5, !!sym(xclmn) + 0.5)),
      !!sym(yclmn) := list(rep(!!sym(yclmn), 2))
    ) %>%
    ungroup() %>%
    unnest(all_of(c(xclmn, yclmn)))
  
  res
}

# Calculate mean signal for pause metaplots
calc_mean_pause_signal <- function(df_in, grps, key_vec = sam_lnms, rel = FALSE) {
  res <- df_in %>%
    group_by(!!!syms(grps)) %>%
    mutate(
      orig_n = n_lab,
      n_lab  = str_c(comma(n_distinct(name)), " pauses\n"),
      n_lab  = str_c(n_lab, comma(n_distinct(gene_name)), " genes"),
    ) %>%
    ungroup()
  
  if (!identical(res$n_lab, res$orig_n)) {
    warning("The number of genes and/or pauses per group is different from the original calculation")
  }
  
  res <- res %>%  
    calc_mean_signal(
      win_col = "win_id",
      rel     = rel,
      grps    = c(grps, "n_lab")
    ) %>%
    format_sample_names(key_vec = key_vec) %>%
    convert_to_bars() %>%
    mutate(win_dist = win_id - 50)
  
  res
}

# Plot mean signal
create_pause_meta <- function(df_in, x, y, color = NULL, plot_clrs = NULL, size = 1, v_line = 0,
                              vline_clr = "black", vline_type = 2, n_lab = NULL, n_lab_pos = c(Inf, Inf),
                              x_lab = waiver(), y_lab = waiver(), ...) {
  res <- df_in
  
  if (!is.null(names(plot_clrs)) && !is.null(color)) {
    res <- res %>%
      mutate(!!sym(color) := fct_relevel(!!sym(color), names(plot_clrs)))
  }
    
  res <- res %>%
    ggplot(aes(!!sym(x), !!sym(y), color = !!sym(color), fill = !!sym(color))) +
    geom_ribbon(
      aes(ymin = 0, ymax = !!sym(y)),
      alpha        = 0.2,
      size         = 0.5,
      outline.type = "upper",
      key_glyph    = draw_key_point,
      ...
    ) +
    guides(fill = guide_legend(override.aes = list(size = 3, alpha = 1))) +
    geom_vline(xintercept = v_line, linetype = vline_type, color = vline_clr) +
    scale_color_manual(values = plot_clrs) +
    scale_fill_manual(values = plot_clrs) +
    theme_info +
    theme(
      legend.position = "none",
      axis.line       = element_blank(),
      panel.border    = element_rect(fill = NA, color = "black")
    ) +
    labs(x = x_lab, y = y_lab)
  
  if (!is.null(n_lab)) {
    res <- res +
      geom_text(
        mapping       = aes(x = n_lab_pos[1], y = n_lab_pos[2], label = !!sym(n_lab)),
        color         = "black",
        size          = txt_pt / 1.2 / .pt,
        alpha         = 1,
        check_overlap = TRUE,
        hjust         = 1.1,
        vjust         = 1.15
      )
  }
  
  res
}

# Create pause logo
create_pause_logo <- function(df_in, left = 9, right = 10, vline = 10.5) {
  
  cen <- vline - 0.5
  
  n_lab <- unique(df_in$n_lab)
  
  res <- df_in %>%
    bed_slop(left = left, right = right, genome, strand = TRUE) %>%
    rename(chr = chrom) %>%
    select(chr, start, end, name, score, strand)
  
  res <- res %>% 
    as.data.frame() %>%
    get.fasta(
      fasta   = fa,
      strand  = TRUE,  # must include all 6 columns for strand option to work
      verbose = FALSE
    ) %>%
    .$sequence
    
  res <- res %>%
    ggseqlogo(seq_type = "dna") +
    
    geom_vline(xintercept = vline, linetype = 2) +
    
    scale_x_continuous(
      labels = (function(x) ifelse(x <= cen, x - (cen + 1), str_c("+", x - cen))),
      breaks = c(1, 6, 15, 20)
    ) +
    
    scale_y_continuous(breaks = breaks_limits) +
    theme_info +
    theme(
      axis.line = element_blank(),
      panel.border = element_rect(fill = NA, color = "black")
    )
  
  res
}

# Create figure showing pause signal and pause logo
create_pause_logo_fig <- function(met_df, pause_df, ttl, y_lab, y_lim = NULL) {
  
  met <- met_df %>%
    create_pause_meta(
      x         = "win_dist",
      y         = "counts",
      color     = "treat",
      v_line    = 0.5,
      plot_clrs = clrs,
      n_lab     = "n_lab",
      x_lab     = "distance from pause (bp)",
      y_lab     = y_lab
    ) +
    coord_cartesian(ylim = y_lim) +
    scale_y_continuous(breaks = breaks_limits) +
    ggtitle(ttl) +
    theme(
      plot.title   = element_text(size = ttl_pt),
      plot.margin  = margin(7, 7, 0, 7, "pt"),
      axis.ticks.x = element_blank(),
      axis.text.x  = element_blank(),
      axis.title.x = element_blank()
    )
  
  lo <- pause_df %>%
    create_pause_logo() +
    theme(plot.margin = margin(0, 7, 7, 7, "pt"))
  
  res <- plot_grid(
    met, lo,
    nrow        = 2,
    align       = "v",
    rel_heights = c(1, 0.5)
  )
  
  res
}

```

```{r "{{grp}} pause beds"}

# Gene regions
# named list with data.frames to intersect pauses with
regs <- c(
  TSS  = "GRCh38_5_TSS_+500bp.bed.gz",
  body = "GRCh38_4_+500bp_pAS.bed.gz",
  pAS  = "GRCh38_3_pAS_+5kb.bed.gz"
)

regs <- regs %>%
  map_chr(~ str_c(list_dir, "/gene_regions/", .x)) %>%
  map(vroom, col_names = bed_cols)

p_dir <- sub_pause_dir$gene

# All pauses
pauses <- load_pause_beds(
  prfxs  = names(p_dir),
  sfxs   = str_c(pause_prfx, "pauses.bed.gz"),
  paths  = unname(p_dir),
  genes  = g_p,
  region = regs
) %>%
  format_sample_names(key_vec = sub_sam_lnms$gene)

# Separated pauses
# Do not format sample names until after merging with metaplot files
sep_pauses <- load_pause_beds(
  prfxs  = names(p_dir),
  sfxs   = str_c(sep_prfx, "pauses.bed.gz"),
  paths  = unname(p_dir),
  genes  = g_p,
  region = regs
)

n_regs <- n_distinct(sep_pauses$region)

```

The number of pause sites shared between samples and replicates is shown below for the TSS (TSS - +500 bp) and gene body (+500 bp - pAS) regions.

* genes >2 kb long and separated by >5 kb
* pauses were identified after downsampling entire gene region

```{r "{{grp}} pause overlaps"}

# Create table of input parameters
nms <- bind_rows(
  expand_grid(treat = treats, region = names(regs)),
  expand_grid(rep   = reps,   region = names(regs))
)

pause_df <- pauses %>%
  nest(pauses = -c(sample, treat, rep, region))

# Create table with overlap for each comparison
pause_overlap <- nms %>%
  pmap_dfr(~ {
    filt_args <- list(...)
    filt_args <- filt_args[!is.na(filt_args)]
    
    # Filter data based on input arguments
    dat <- pause_df
    
    filt_args %>%
      iwalk(~ {
        dat <<- dat %>%
          filter(!!sym(.y) == .x)
      })
    
    if (nrow(dat) < 2) {
      return(NULL)
    }
    
    # Calculate overlap
    dfs <- set_names(dat$pauses, dat$sample)
    res <- calc_pause_overlap(dfs)
    
    res <- dat %>%
      full_join(res, by = c(sample = "key")) %>%
      mutate(comp = str_c(names(dfs), collapse = " "))
    
    # Fill in NAs
    res <- res %>%
      mutate(across(everything(), ~ {
        if (n_distinct(.x, na.rm = TRUE) == 1) {
          replace_na(.x, unique(na.omit(.x)))
          
        } else {
          .x
        }
      }))
        
    res
  })

n_comps <- n_distinct(pause_overlap$comp)

# Print pause stats
pause_overlap %>%
  filter(class == "unique") %>%
  distinct(sample, region, n_genes, n_pauses) %>%
  group_by(sample) %>%
  mutate(total_pauses = sum(n_pauses)) %>%
  ungroup()

```

```{r "{{grp}} pause overlap bars", fig.width = n_comps * 3, fig.height = 4}

# Bar graph colors
clr_df <- sam_df %>%
  separate(sample, sep = "_", into = c("treat", "rep"), remove = FALSE) %>%
  mutate(
    rep_i = as.numeric(str_extract(rep, "[0-9]+$")),
    clr   = clrs[treat],
    clr   = lighten(clr, 0.3 * (rep_i - 1))
  )

bar_clrs <- set_names(clr_df$clr, clr_df$sample)
bar_clrs <- c(shared = "grey85", bar_clrs)

# Bar graph data
plt_dat <- pause_overlap %>%
  mutate(
    sample = ifelse(class == "shared", class, sample),
    sample = fct_relevel(sample, names(bar_clrs)),
    comp   = fct_inorder(comp),
    region = fct_relevel(region, names(regs))
  ) %>%
  group_by(comp, region) %>%
  mutate(
    comp_reg = str_c(comp, "_", region),
    x_lab    = str_c(comma(sum(n)), "\npauses")
  ) %>%
  ungroup()

# Fill labels for legend
plt_dat <- plt_dat %>%
  group_by(comp) %>%
  mutate(
    fill_lab = case_when(
      n_distinct(treat, na.rm = TRUE) == 1 ~ rep,
      n_distinct(rep,   na.rm = TRUE) == 1 ~ treat
    ),
    fill_lab = if_else(
      is.na(fill_lab),
      as.character(sample),
      fill_lab
    )
  ) %>%
  ungroup()

# X-axis labels
x_labs <- distinct(plt_dat, comp_reg, x_lab)

x_labs <- set_names(
  x_labs$x_lab,
  x_labs$comp_reg
)

# Split data.frame into list for plotting
df_lst <- plt_dat %>%
  arrange(region) %>%
  mutate(comp_reg = fct_inorder(comp_reg)) %>%
  group_by(comp) %>%
  group_split()
  
# Create bar graphs
bars <- df_lst %>%
  map(~ {
    fill_labs <- set_names(
      unique(.x$fill_lab),
      unique(.x$sample)
    )
    
    plt_clrs <- bar_clrs[names(bar_clrs) %in% .x$sample]
    
    .x %>%
      mutate(
        sample = fct_drop(sample),
        sample = fct_relevel(sample, "shared", after = 1)
      ) %>%
      ggplot(aes(comp_reg, n, fill = sample)) +
      geom_col(position = "fill", width = 0.95) +
      scale_fill_manual(values = plt_clrs, labels = fill_labs) +
      scale_x_discrete(labels = x_labs) +
      facet_wrap(~ region, scales = "free") +

      ggtitle(unique(.x$comp)) +
      theme_info +
      theme(
        aspect.ratio     = 3,
        strip.text       = element_text(vjust = 0),
        legend.position  = "left",
        legend.title     = element_blank(),
        panel.border     = element_blank(),
        panel.spacing    = unit(-10, "pt"),
        axis.title       = element_blank(),
        axis.line        = element_blank(),
        axis.ticks       = element_blank(),
        axis.text.x      = element_text(vjust = 4),
        axis.text.y      = element_blank()
      )
  })

bars <- plot_grid(
  plotlist = bars,
  nrow     = 1
)

bars

```

<br>

### Mean signal

Mean mNET-seq signal is shown below for pause sites identified for TSS (TSS - +500 bp) and gene body (+500 bp - pAS) regions. The sequence preference for each group of sites is shown below each plot. The dotted line indicates the 3' end of the nascent RNA.

* genes >2 kb long and separated by >5 kb
* pauses were identified after downsampling entire gene region
* pauses separated by >30 bp

```{r "{{grp}} pauses mean"}

# Load and merge bed files
sfxs <- c("sense" = str_c(pause_prfx, "pauses_meta_N.bed.gz"))
out  <- here(params$obj_dir, str_c(grp, str_c(pause_prfx, "pauses_meta.tsv.gz")))

merge_p <- load_merge_wins(
  prfxs         = names(p_dir),      # Sample names
  sfxs          = sfxs,              # Suffix list with types as names
  paths         = p_dir,             # Directories containing files
  group         = grp,               # Group name
  
  file_out      = out,               # Path to output file
  overwrite     = params$overwrite,  # Overwrite output file if it exists
  col_names     = win_cols,          # Column names for bed files
  filter_unique = FALSE,             # Remove genes that are not shared between all samples
  win_num       = 100,               # Total number of expected windows (including sense + anti)
  win_min       = 41,
  win_max       = 60,
  ref_win       = NULL               # Reference window for adjusting window ids
)

# Filter pauses for separated genes
# add gene regions
merge_p <- sep_pauses %>%
  select(name, gene_name, sample, region, n_lab) %>%
  inner_join(merge_p, by = c("name", "sample"))

# Filter for overlapping genes
# want all samples for each region to use the same genes
# need to recalculate n values since filtering genes/pauses
merge_p <- merge_p %>%
  group_by(region) %>%
  group_split() %>%
  map_dfr(
    merge_wins,
    groups        = "sample",
    ref_win       = NULL,
    filter_unique = TRUE,
    by            = "gene_name"
  ) %>%
  group_by(sample, region) %>%
  mutate(
    n_lab = comma(n_distinct(name)),
    n_lab = str_c(n_lab, " pauses\n"),
    n_lab = str_c(n_lab, comma(n_distinct(gene_name))),
    n_lab = str_c(n_lab, " genes")
  ) %>%
  ungroup()

merge_clmns <- c("name", "gene_name", "sample", "region")

filt_pauses <- merge_p %>%
  distinct(!!!syms(merge_clmns), n_lab)

# filter pauses so they match merge_p
filt_pauses <- sep_pauses %>%
  select(-n_lab) %>%
  inner_join(filt_pauses, by = merge_clmns) %>%
  group_by(sample, region) %>%
  mutate(
    orig_lab = n_lab,
    n_lab    = comma(n_distinct(name)),
    n_lab    = str_c(n_lab, " pauses\n"),
    n_lab    = str_c(n_lab, comma(n_distinct(gene_name))),
    n_lab    = str_c(n_lab, " genes")
  ) %>%
  ungroup() %>%
  format_sample_names(key_vec = sub_sam_lnms$gene)

stopifnot(all(filt_pauses$n_lab == filt_pauses$orig_lab))

filt_pauses <- filt_pauses %>%
  select(-orig_lab)

# Calculate mean signal
mean_grps <- c("sample", "group", "type", "region")

mean_p <- merge_p %>%
  calc_mean_pause_signal(
    mean_grps,
    key_vec = sub_sam_lnms$gene,
    rel     = FALSE
  )

rel_p <- merge_p %>%
  calc_mean_pause_signal(
    mean_grps,
    key_vec = sub_sam_lnms$gene,
    rel     = TRUE
  )

# Create nested data.frame with data for metaplots and logos
p_dat <- list(
  "mean_counts" = mean_p,
  "rel_counts"  = rel_p,
  "pauses"      = filt_pauses
)

sep_pause_df <- p_dat %>%
  imap(~ {
    .x %>%
      nest(!!sym(.y) := -c(sample, region))
  }) %>%
  reduce(full_join, by = c("sample", "region")) %>%
  separate(sample, sep = "_", into = c("treat", "rep"), remove = FALSE)

# Check for rows not shared between all data.frames
if (!all(complete.cases(select_if(sep_pause_df, is.character)))) {
  stop("Not all rows shared, malformed data.frame")
}

# Check that n_lab matches for all data.frames
n_chk <- sep_pause_df %>%
  mutate(
    across(
      all_of(names(p_dat)),
      ~ map_chr(.x, ~ unique(.x$n_lab))
    )
  ) %>%
  pivot_longer(all_of(names(p_dat))) %>%
  group_by(sample, region) %>%
  filter(length(unique(value)) > 1)

stopifnot(nrow(n_chk) == 0)

```

```{r "{{grp}} pauses mean meta", include = FALSE}

# Plot mean signal with logos
y_lab    <- "RPKM"
y_lim    <- range(mean_p$counts)
y_lim[1] <- 0

mean_figs <- sep_pause_df %>%
  mutate(
    ttl = str_c(treat, rep, region, sep = " "),
    fig = pmap(
      list(mean_counts, pauses, ttl),
      create_pause_logo_fig,
      y_lab = y_lab,
      y_lim = y_lim
    ),
    sample = fct_relevel(sample, sam_nms),
    region = fct_relevel(region, names(regs))
  ) %>%
  arrange(region, sample)

```

```{r "{{grp}} pauses mean meta 2", fig.width = 3 * n_sams, fig.height = 3.2 * n_regs}

# Run this in separate chunk to exclude fasta messages
plot_grid(
  plotlist = mean_figs$fig,
  ncol     = n_sams
)

```

<br>

### Relative signal

Mean relative mNET-seq signal is shown below for pause sites identified for TSS (TSS - +500 bp), gene body (+500 bp - pAS), and pAS (pAS - +5 kb) regions. The sequence preference for each group of sites is shown below each plot. The dotted line indicates the 3' end of the nascent RNA.

* genes >2 kb long and separated by >5 kb
* pauses separated by >30 bp

```{r "{{grp}} pauses rel meta", include = FALSE}

y_lab    <- "relative signal"
y_lim    <- range(rel_p$counts)
y_lim[1] <- 0

rel_figs <- sep_pause_df %>%
  mutate(
    ttl = str_c(sample, " ", region),
    fig = pmap(
      list(rel_counts, pauses, ttl),
      create_pause_logo_fig,
      y_lab = y_lab,
      y_lim = y_lim
    ),
    sample = fct_relevel(sample, sam_nms),
    region = fct_relevel(region, names(regs))
  ) %>%
  arrange(region, sample)

```

```{r "{{grp}} pauses rel meta 2", fig.width = 3 * n_sams, fig.height = 3.2 * n_regs}

# Run this in separate chunk to exclude fasta messages
plot_grid(
  plotlist = rel_figs$fig,
  ncol     = n_sams
)

```
